### ISSUE TITLE
Allow working without Docker

### ISSUE THREAD

**Issue by gnn** [openego/eGon-data#64]
Since there are circumstances which might prevent users from using Docker, we should have a way of not using Docker and instead use PostgreSQL installation on the local system. The best way would probably be to allow customizing the configuration for the local database, checking whether there's already a service listening on the specified port and then deciding whether to start a docker container or not.
The bigger problem which this entails is that, since there will usually not be a clean slate on the local database, most import tasks will either have to clean up everything before importing their stuff, or checking whether their import is already done before starting their work.

**Comment by gplssm** [comment]
> Since there are circumstances which might prevent users from using Docker, we should have a way of not using Docker and instead use PostgreSQL installation on the local system. The best way would probably be to allow customizing the configuration for the local database, checking whether there's already a service listening on the specified port and then deciding whether to start a docker container or not.

In general the idea of checking if a service is listening on the desired port is valuable, in particular **with** using a dockered database. This could be done in `initdb()` and call of `docker-compose ...` might be skipped.
But, it does not help us to decide on spinning up the docker container or not in case we want to use the locally installed database.

> The bigger problem which this entails is that, since there will usually not be a clean slate on the local database, most import tasks will either have to clean up everything before importing their stuff, or checking whether their import is already done before starting their work.

This opens up other, bigger questions!
_open side discussion_
Right now, data imports (at these that are implement by myself) overwrite data without checking anything. The idea behind this was to decide elsewhere if the data should be imported to the datase, aka. if the related task should be executed. The reason for that is, that I would like to have functionality that checks if the data which might exist in the database is of the same version as the data that would result from executing the related script(s). This should be implemented for the entire workflow.
_end of side discussion_ 

With the above in mind, I would suggest to implement a switch at prominent place to actively choose to use the locally installed database. Preferable with something like

`egon-data serve --local-database`

As this command currently calls `initdb()` it might be possible to forward this flag, such that docker isn't called but it is checked the desired database exists in the local postgres installation and otherwise it gets created. With doing so, we could simply advice to use the cli accordingly and then work in the airflow UI.

Does this sound reasonable?

**Comment by gnn** [comment]
> > Since there are circumstances which might prevent users from using Docker, we should have a way of not using Docker and instead use PostgreSQL installation on the local system. The best way would probably be to allow customizing the configuration for the local database, checking whether there's already a service listening on the specified port and then deciding whether to start a docker container or not.
>
> In general the idea of checking if a service is listening on the desired port is valuable, in particular **with** using a dockered database. This could be done in `initdb()` and call of `docker-compose ...` might be skipped.

I agree, although IIRC, the `docker-compose` call just does nothing if the container is already up, so this is just an optimization. But please correct me if I'm wrong.


> But, it does not help us to decide on spinning up the docker container or not in case we want to use the locally installed database.

I think it does. We should be agnostic about whether we are using a docker container or a local database. Both are available on "localhost". The user just configures the port used to connect to the database he wants to use with the workflow. If there is a service listening on this port, then we don't start the docker container. At least that's how I envision it.


> > The bigger problem which this entails is that, since there will usually not be a clean slate on the local database, most import tasks will either have to clean up everything before importing their stuff, or checking whether their import is already done before starting their work.

Note that I was wrong on this. I wrote this sentence with the impression, that changes to the database in the container do not persist between reboots of the container. I now know that this is wrong, because the database data is stored outside of the container on the user's machine. So the issue of having to continuously update existing data instead of always rebuilding the database from scratch is independent of whether to use a locally installed database or on inside a container. My bad.


> This opens up other, bigger questions!
> _open side discussion_
> Right now, data imports (at these that are implement by myself) overwrite data without checking anything. The idea behind this was to decide elsewhere if the data should be imported to the datase, aka. if the related task should be executed. The reason for that is, that I would like to have functionality that checks if the data which might exist in the database is of the same version as the data that would result from executing the related script(s). This should be implemented for the entire workflow.
> _end of side discussion_

This has more than one side. On the one hand, having a general infrastructure available for checking database data against data to be imported is desirable. On the other hand the logic to do this kind of checking will pretty surely differ from dataset to dataset.
Should we open a different issue to discuss how to proceed with implementing this feature? I actually have an idea of how to do it for the current import of the OSM dataset.


> With the above in mind, I would suggest to implement a switch at prominent place to actively choose to use the locally installed database. Preferable with something like
>
> `egon-data serve --local-database`
>
> As this command currently calls `initdb()` it might be possible to forward this flag, such that docker isn't called but it is checked the desired database exists in the local postgres installation and otherwise it gets created. With doing so, we could simply advice to use the cli accordingly and then work in the airflow UI.
>
> Does this sound reasonable?

As I stated in the beginning, I think we should do things a bit differently. There should be a command line switch specifying a different configuration file than the one shipped with the package and there should also be switches overriding individual configuration options. Then checking for available services solves the problem.

**Comment by gplssm** [comment]
> > > Since there are circumstances which might prevent users from using Docker, we should have a way of not using Docker and instead use PostgreSQL installation on the local system. The best way would probably be to allow customizing the configuration for the local database, checking whether there's already a service listening on the specified port and then deciding whether to start a docker container or not.
> > 
> > 
> > In general the idea of checking if a service is listening on the desired port is valuable, in particular **with** using a dockered database. This could be done in `initdb()` and call of `docker-compose ...` might be skipped.
> 
> I agree, although IIRC, the `docker-compose` call just does nothing if the container is already up, so this is just an optimization. But please correct me if I'm wrong.

For me it seems that the container is re-created. The task takes 17 s in the workflow...

> 
> > But, it does not help us to decide on spinning up the docker container or not in case we want to use the locally installed database.
> 
> I think it does. We should be agnostic about whether we are using a docker container or a local database. 

Check!

> Both are available on "localhost". The user just configures the port used to connect to the database he wants to use with the workflow. If there is a service listening on this port, then we don't start the docker container. At least that's how I envision it.
> 

If it's easier for us, we could also expect from users with a local postgres installation that they use the same configuration as the dockered database.

> > > The bigger problem which this entails is that, since there will usually not be a clean slate on the local database, most import tasks will either have to clean up everything before importing their stuff, or checking whether their import is already done before starting their work.
> 
> Note that I was wrong on this. I wrote this sentence with the impression, that changes to the database in the container do not persist between reboots of the container. I now know that this is wrong, because the database data is stored outside of the container on the user's machine. So the issue of having to continuously update existing data instead of always rebuilding the database from scratch is independent of whether to use a locally installed database or on inside a container. My bad.

To remove persistence, use flag `-rm` for `docker-compose`

> > This opens up other, bigger questions!
> > _open side discussion_
> > Right now, data imports (at these that are implement by myself) overwrite data without checking anything. The idea behind this was to decide elsewhere if the data should be imported to the datase, aka. if the related task should be executed. The reason for that is, that I would like to have functionality that checks if the data which might exist in the database is of the same version as the data that would result from executing the related script(s). This should be implemented for the entire workflow.
> > _end of side discussion_
> 
> This has more than one side. On the one hand, having a general infrastructure available for checking database data against data to be imported is desirable. On the other hand the logic to do this kind of checking will pretty surely differ from dataset to dataset.
> Should we open a different issue to discuss how to proceed with implementing this feature? I actually have an idea of how to do it for the current import of the OSM dataset.

Yes. And I would prefer a discussion in a call.

> > With the above in mind, I would suggest to implement a switch at prominent place to actively choose to use the locally installed database. Preferable with something like
> > `egon-data serve --local-database`
> > As this command currently calls `initdb()` it might be possible to forward this flag, such that docker isn't called but it is checked the desired database exists in the local postgres installation and otherwise it gets created. With doing so, we could simply advice to use the cli accordingly and then work in the airflow UI.
> > Does this sound reasonable?
> 
> As I stated in the beginning, I think we should do things a bit differently. There should be a command line switch specifying a different configuration file than the one shipped with the package and there should also be switches overriding individual configuration options. Then checking for available services solves the problem.

I'm totally open to implement it via different config file. I would just like to make it not to complicated in order to provide this feature soon.

Can you make a proposal of how you think it should be implemented. I can help implementing it.
